src/query/storage/promremote/storage.go

```diff
@@ -159,26 +158,25 @@ func NewStorage(opts Options) (storage.Storage, error) {
 		client:          client,
 		endpointMetrics: initEndpointMetrics(opts.endpoints, scope),
 		scope:           scope,
-		samplesWritten:  scope.Counter("samples_written"),
+		enqueuedSamples: scope.Counter("enqueued_samples"),
+		writtenSamples:  scope.Counter("written_samples"),
+		droppedSamples:  scope.Counter("dropped_samples"),
+		inFlightSamples: scope.Gauge("in_flight_samples"),
+		batchWrites:     scope.Counter("batch_writes"),
+		tickWrites:      scope.Counter("tick_writes"),
 		droppedWrites:   scope.Counter("dropped_writes"),
+		errWrites:       scope.Counter("err_writes"),
+		retryWrites:     scope.Counter("retry_writes"),
 		dupWrites:       scope.Counter("duplicate_writes"),
-		ingestorWrites:  scope.Counter("ingestor_writes"),
-		enqueued:        scope.Counter("enqueued"),
-		enqueueErr:      scope.Counter("enqueue_error"),
-		batchWrite:      scope.Counter("batch_write"),
-		batchWriteErr:   scope.Counter("batch_write_err"),
-		tickWrite:       scope.Counter("tick_write"),
-		retryWrite:      scope.Counter("retry_write"),
 		logger:          opts.logger,
-		queryQueue:      make(chan *storage.WriteQuery, opts.queueSize),
-		workerPool:      xsync.NewWorkerPool(opts.poolSize),
-		pendingQuery:    queriesWithFixedTenants,
-		noTenantFound:   scope.Counter("no_tenant_found"),
-		cancel:          cancel,
-		writeLoopDone:   make(chan struct{}),
-		wrongTenant:     scope.Counter("wrong_tenant"),
+		// this queue is used for enqueueing all data, need to be larger to avoid data loss
+		// this can be tested in TestLoad
+		queryQueue:    make(chan *storage.WriteQuery, len(opts.tenantRules)*opts.queueSize),
```

Comment:
Increase the queryQueue size will pass the unit tests and mitigate the chance of data samples not being proceeded fast and loss, we can also add a metric to monitoring the queue size.

Reply:
This should be fine. Not too much added memory. The for loop should drain the channel until all values are read. Can you write a test to make sure writes in the channel are also flushed (not just in the per-tenant queue)?
<img width="646" alt="image" src="https://github.com/databricks/m3/assets/129140017/2450bd37-4cd6-4e46-9ca5-bedcfdad0600">

<img width="659" alt="image" src="https://github.com/databricks/m3/assets/129140017/3bfe8ed1-1868-40b3-a727-a04af1a20366">